pkg/reconciler/workload/apply.go

```diff
@@ -40,52 +36,41 @@ var (
 	errPatchKubeApp          = "cannot patch existing KubernetesApplication"
 )
 
-// KubeAppApply applies changes to the supplied KubernetesApplication. The
-// KubernetesApplication will be created if it does not exist, or patched if it
-// does. Resource templates are merged instead of replaced if they have the same
-// name and GroupVersionKind.
-func KubeAppApply(ctx context.Context, c client.Client, o runtime.Object, ao ...resource.ApplyOption) error {
-	opts := &resource.ApplyOptions{}
-	for _, fn := range ao {
-		fn(opts)
-	}
-
-	m, ok := o.(*workloadv1alpha1.KubernetesApplication)
-	if !ok {
-		return errors.New(errNotKubeApp)
-	}
-
-	desired := &workloadv1alpha1.KubernetesApplication{}
-	m.DeepCopyInto(desired)
-
-	err := c.Get(ctx, types.NamespacedName{Name: m.GetName(), Namespace: m.GetNamespace()}, o)
-	if kerrors.IsNotFound(err) {
-		return errors.Wrap(c.Create(ctx, o), errCreateKubeApp)
-	}
-	if err != nil {
-		return errors.Wrap(err, errGetKubeApp)
-	}
-
-	if opts.ControllersMustMatch && !meta.HaveSameController(m, desired) {
-		return errors.New(errDiffController)
-	}
-
-	for i, temp := range m.Spec.ResourceTemplates {
-		if desired.Spec.ResourceTemplates[i].Spec.Template.GroupVersionKind() != temp.Spec.Template.GroupVersionKind() {
-			continue
+// KubeAppApplyOption ensures resource templates are merged instead of replaced
+// before patch if they have the same name and GroupVersionKind.
```

Comment:
It might be worth expanding on this docstring to point out what this apply option does why we're doing it (i.e. to work around the lack of array merged in merge patches). It took me a few reads to grok how it all works, even with context on the problem. I'd especially call attention to the fact that any template that isn't in our desired list will be removed automatically by `resource.Apply` (since we'll overwrite the entire array using our merge patch).